File: GCDLib/Core/GCDProject/ProjectManagerUI.vb
Imports System.IO
Imports System.Windows.Forms

Namespace Core.GCDProject

    Public Class ProjectManagerUI
        Inherits ProjectManagerBase

        Protected Shared m_PyramidManager As RasterPyramidManager
        Private Shared m_FISLibrary As FileInfo

        Public Shared ReadOnly Property IsArcMap As Boolean
            Get
                Return Reflection.Assembly.GetEntryAssembly().FullName.ToLower().Contains("arcmap")
            End Get
        End Property

        Public Shared ReadOnly Property PyramidManager As RasterPyramidManager
            Get
                Return m_PyramidManager
            End Get
        End Property

        Public Shared Property FISLibrary As List(Of ErrorCalculation.FIS.FISLibraryItem)
            Get
                If m_FISLibrary.Exists Then
                    Return ErrorCalculation.FIS.FISLibraryItem.Load(m_FISLibrary)
                Else
                    Return New List(Of ErrorCalculation.FIS.FISLibraryItem)
                End If
            End Get
            Set(value As List(Of ErrorCalculation.FIS.FISLibraryItem))
                ErrorCalculation.FIS.FISLibraryItem.Save(m_FISLibrary, value)
            End Set
        End Property

        Private Shared ReadOnly Property ApplicationFolder As String
            Get
                Return Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "GCD")
            End Get
        End Property

        Public Sub New(ByVal eDefaultRasterType As GCDConsoleLib.Raster.RasterDriver, ByVal sAutomaticPyramids As String)
            MyBase.New(ApplicationFolder, My.Settings.Erosion, My.Settings.Depsoition, eDefaultRasterType)

            m_PyramidManager = New RasterPyramidManager(sAutomaticPyramids)
            m_FISLibrary = New FileInfo(Path.Combine(Environment.SpecialFolder.ApplicationData, "FISLibrary.xml"))

        End Sub

        ''' <summary>
        ''' 
        ''' </summary>
        ''' <param name="bRemoveMissingrasters"></param>
        ''' <returns>True if the GCD project has been altered during the validation process</returns>
        Public Shared Function Validate(Optional bRemoveMissingrasters As Boolean = False) As Boolean

            If Not My.Settings.ValidateProjectOnLoad Then
                Return False
            End If

            Dim bEditedGCDFile As Boolean = False
            Dim sOutputMessages As New Text.StringBuilder

            Dim rProject As ProjectDS.ProjectRow = CurrentProject
            If Not TypeOf rProject Is ProjectDS.ProjectRow Then
                Return bEditedGCDFile
            End If

            Dim lItemsToRemove As New List(Of ProjectDS.DEMSurveyRow)
            For Each rDEM As ProjectDS.DEMSurveyRow In rProject.GetDEMSurveyRows
                Dim sPath As String = GetAbsolutePath(rDEM.Source)
                If Not GCDConsoleLib.GISDataset.FileExists(sPath) Then
                    lItemsToRemove.Add(rDEM)
                End If
            Next

            'Pass bEditedGCDFile as a reference argument between these 3 methods 
            bEditedGCDFile = ValidateAssociatedSurfaces(sOutputMessages, bEditedGCDFile, bRemoveMissingrasters)
            bEditedGCDFile = ValidateErrorSurfaces(sOutputMessages, bEditedGCDFile, bRemoveMissingrasters)
            bEditedGCDFile = ValidateDoDAnalysisSurfaces(rProject, sOutputMessages, bEditedGCDFile, bRemoveMissingrasters)

            If bRemoveMissingrasters Then

                CheckCopyStatusOfGCDProject(FilePath, bEditedGCDFile, lItemsToRemove.Count)

                For Each rDEM As ProjectDS.DEMSurveyRow In lItemsToRemove
                    sOutputMessages.Append("Removing DEM Survey '" & rDEM.Name & "' because the source raster is missing or cannot be found.").AppendLine()
                    ds.DEMSurvey.RemoveDEMSurveyRow(rDEM)
                    Debug.Write(sOutputMessages.ToString)
                    bEditedGCDFile = True
                Next
            End If

            Try
                save()

            Catch ex As Exception
                ex.Data("Removing datasets") = sOutputMessages
                Throw
            End Try

            Return bEditedGCDFile

            ' This code cleans the project of empty folders which were not fully deleted due to inability to remove
            ' ESRI locks even after the file has successfully been deleted And removed from GCD xml
            CleanProjectFolderStructure(True)

            Return bEditedGCDFile

        End Function

        Private Shared Function ValidateAssociatedSurfaces(ByRef sOutputMessages As Text.StringBuilder, ByRef bEditedGCDFile As Boolean, Optional ByVal bRemoveMissingRasters As Boolean = False) As Boolean

            Dim lItemsToRemove As New List(Of ProjectDS.AssociatedSurfaceRow)
            For Each rAssoc As ProjectDS.AssociatedSurfaceRow In ds.AssociatedSurface.Rows
                Dim dbTest = CType(rAssoc, DataRow)
                If IsDBNull(dbTest("OriginalSource")) Then
                    rAssoc.OriginalSource = rAssoc.Source
                    Dim sAssociatedSurfaceName As String = System.IO.Path.GetFileNameWithoutExtension(rAssoc.OriginalSource)
                    rAssoc.Source = GetRelativePath(m_OutputManager.AssociatedSurfaceRasterPath(rAssoc.DEMSurveyRow.Name, sAssociatedSurfaceName))
                    CheckCopyStatusOfGCDProject(FilePath, bEditedGCDFile, 1)
                    bEditedGCDFile = True
                    save()
                End If
                Dim sPath As String = GetAbsolutePath(rAssoc.Source)
                If Not GCDConsoleLib.GISDataset.FileExists(sPath) Then
                    lItemsToRemove.Add(rAssoc)
                End If
            Next

            If bRemoveMissingRasters Then

                'Check to see if a copy of the .gcd project is necessary
                CheckCopyStatusOfGCDProject(FilePath, bEditedGCDFile, lItemsToRemove.Count)

                For Each rAssoc As ProjectDS.AssociatedSurfaceRow In lItemsToRemove
                    sOutputMessages.Append("Removing Associated Surface '" & rAssoc.Name & "' because the source raster is missing or cannot be found.").AppendLine()
                    ds.AssociatedSurface.RemoveAssociatedSurfaceRow(rAssoc)
                    Debug.Write(sOutputMessages.ToString)
                    bEditedGCDFile = True
                Next
            End If

            Return bEditedGCDFile

        End Function

        Private Shared Function ValidateErrorSurfaces(ByRef sOutputMessages As Text.StringBuilder, ByRef bEditedGCDFile As Boolean, Optional ByVal bRemoveMissingRasters As Boolean = False) As Boolean

            Dim lItemsToRemove As New List(Of ProjectDS.ErrorSurfaceRow)
            For Each rError As ProjectDS.ErrorSurfaceRow In ds.ErrorSurface.Rows
                If rError.IsSourceNull Then
                    lItemsToRemove.Add(rError)
                Else
                    Dim sPath As String = GetAbsolutePath(rError.Source)
                    If Not GCDConsoleLib.GISDataset.FileExists(sPath) Then
                        lItemsToRemove.Add(rError)
                    End If
                End If
            Next

            If bRemoveMissingRasters Then

                CheckCopyStatusOfGCDProject(FilePath, bEditedGCDFile, lItemsToRemove.Count)

                For Each rError As ProjectDS.ErrorSurfaceRow In lItemsToRemove
                    Dim sMessage As String = "Removing Error Surface '" & rError.Name & "' because the source raster is missing or cannot be found."
                    ds.ErrorSurface.RemoveErrorSurfaceRow(rError)
                    sOutputMessages.Append(sMessage).AppendLine()
                    Debug.Write(sOutputMessages.ToString)
                    bEditedGCDFile = True
                Next
            End If

            Return bEditedGCDFile

        End Function

        Private Shared Function ValidateDoDAnalysisSurfaces(ByRef rProjectRow As ProjectDS.ProjectRow, ByRef sOutputMessages As Text.StringBuilder, ByRef bEditedGCDFile As Boolean, Optional ByVal bRemoveMissingRasters As Boolean = False) As Boolean

            Dim lItemsToRemove As New List(Of ProjectDS.DoDsRow)
            For Each rDoD As ProjectDS.DoDsRow In ds.DoDs.Rows
                Dim sPathDoDRaw As String = GetAbsolutePath(rDoD.RawDoDPath)
                Dim sPathDoDThresh As String = GetAbsolutePath(rDoD.ThreshDoDPath)
                If Not GCDConsoleLib.GISDataset.FileExists(sPathDoDRaw) Or Not GCDConsoleLib.GISDataset.FileExists(sPathDoDThresh) Then
                    lItemsToRemove.Add(rDoD)
                End If
            Next

            If bRemoveMissingRasters Then

                CheckCopyStatusOfGCDProject(FilePath, bEditedGCDFile, lItemsToRemove.Count)

                For Each rDoD As ProjectDS.DoDsRow In lItemsToRemove
                    sOutputMessages.Append("Removing DoD Surface '" & rDoD.Name & "' because the source raster is missing or cannot be found.").AppendLine()
                    ds.DoDs.RemoveDoDsRow(rDoD)
                    Debug.Write(sOutputMessages.ToString)
                    bEditedGCDFile = True
                Next
            End If

            Return bEditedGCDFile

        End Function

        Private Shared Sub CleanProjectFolderStructure(Optional ByVal bRemoveEmptyFolders As Boolean = False)

            If FilePath Is Nothing Then
                Exit Sub
            End If

            Dim sPath As String = IO.Path.GetDirectoryName(FilePath)
            Try
                If bRemoveEmptyFolders Then
                    DeleteEmptyFolders(sPath)
                End If
            Catch ex As Exception
                'Do Nothing
            End Try

        End Sub

        Private Shared Function CheckCopyStatusOfGCDProject(ByVal sGCDProjectFilePath As String, ByVal bEditedGCDFile As Boolean, ByVal iItemsToRemove As Integer) As String

            If String.IsNullOrEmpty(sGCDProjectFilePath) Then
                Return Nothing
            End If

            'If gcd file has not been edited in previous steps
            If bEditedGCDFile = False Then
                'If there are elements of GCD file currently marked for removal
                If iItemsToRemove > 0 Then

                    'Have established that gcd file has not been edited, i.e. don't want to make a backup of something that has already changed
                    'and established that there items from this set to be removed from the gcd file therefore make a copy of the .gcd file

                    'Copy GCD file, give backup file timestamp of when copy is made
                    Dim sGCDProjectCopyName As String = System.IO.Path.GetFileNameWithoutExtension(sGCDProjectFilePath)
                    Dim sTimeStamp As DateTime = DateTime.Now
                    Dim format As String = "yyyydMHHmmss" ' format is one integer of year day hours minutes seconds
                    sGCDProjectCopyName = sGCDProjectCopyName & sTimeStamp.ToString(format) & ".gcd"
                    Dim sBackUpPath As String = IO.Path.Combine(System.IO.Path.GetDirectoryName(sGCDProjectFilePath), "Backup")
                    sGCDProjectCopyName = System.IO.Path.Combine(sBackUpPath, sGCDProjectCopyName)

                    If System.IO.Directory.Exists(sBackUpPath) Then
                        System.IO.File.Copy(sGCDProjectFilePath, sGCDProjectCopyName)
                    Else
                        System.IO.Directory.CreateDirectory(sBackUpPath)
                        System.IO.File.Copy(sGCDProjectFilePath, sGCDProjectCopyName)
                    End If

                    Return sGCDProjectCopyName

                End If
            End If

            Return Nothing

        End Function

        Private Shared Sub DeleteEmptyFolders(ByVal sPath As String)

            Dim SubDirectories() As String = Directory.GetDirectories(sPath)
            For Each strDirectory As String In SubDirectories
                DeleteEmptyFolders(strDirectory)
            Next

            If Directory.GetFiles(sPath).Length + Directory.GetDirectories(sPath).Length = 0 Then
                Directory.Delete(sPath)
                Dim sMessage As String = "Deleting folder '" & sPath & "' from project file system because the folder is empty." & vbCrLf
                Debug.Write(sMessage.ToString)
            End If
        End Sub

        Public Shared Sub CopyDeployFolder()

            Dim sDestinationFolder As String = ApplicationFolder

            'New Code to test this may not be what you intend though
            Dim sFolderName As String = "Deploy"
            If String.Compare(sFolderName, "Reources\FIS", True) = 0 Then

                sDestinationFolder = IO.Path.Combine(sDestinationFolder, sFolderName)
                Dim sExecutingAssemblyFolder = System.IO.Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location)
                Dim sOriginalFolder As String = IO.Path.Combine(sExecutingAssemblyFolder, sFolderName)

                If IO.Directory.Exists(IO.Path.GetDirectoryName(sDestinationFolder)) Then
                    CopyDirectory(sOriginalFolder, sDestinationFolder)
                ElseIf Not IO.Directory.Exists(IO.Path.GetDirectoryName(sDestinationFolder)) Then
                    IO.Directory.CreateDirectory(sDestinationFolder)
                    Debug.WriteLine("Copying AddIn Folder """ & sOriginalFolder & """ to """ & sDestinationFolder & """")
                    CopyDirectory(sOriginalFolder, sDestinationFolder)
                End If
            Else
                Directory.CreateDirectory(sDestinationFolder)
                Dim sOriginalFolder As String = IO.Path.Combine(Path.GetDirectoryName(Reflection.Assembly.GetExecutingAssembly().Location), "Deploy")
                'sDestinationFolder = Path.Combine(sDestinationFolder, Path.GetFileNameWithoutExtension(sFolderName))
                Debug.WriteLine("Copying AddIn Folder """ & sOriginalFolder & """ to """ & sDestinationFolder & """")
                CopyDirectory(sOriginalFolder, ApplicationFolder)
            End If

        End Sub

        Private Shared Sub CopyDirectory(ByVal sourcePath As String, ByVal destPath As String)

            If Not IO.Directory.Exists(destPath) Then
                Try
                    Dim dFolder As IO.DirectoryInfo = IO.Directory.CreateDirectory(destPath)
                    If Not dFolder.Exists Then
                        Dim ex As New Exception("Failed to create directory.")
                        ex.Data("Directory") = destPath
                        Throw ex
                    End If
                Catch ex As IO.DirectoryNotFoundException
                    Dim ex2 As New Exception("The specified path is invalid (for example, it is on an unmapped drive).", ex)
                    Throw ex2

                Catch ex As IO.IOException
                    Dim ex2 As New Exception("The directory specified by path is a file or the network name is not known.", ex)
                    Throw ex2

                Catch ex As UnauthorizedAccessException
                    Dim ex2 As New Exception("The caller does not have the required permission.", ex)
                    Throw ex2

                Catch ex As ArgumentNullException
                    Dim ex2 As New Exception("Path is Nothing.", ex)
                    Throw ex2

                Catch ex As ArgumentException
                    Dim ex2 As New Exception("The path is a zero-length string, contains only white space, or contains one or more invalid characters. You can query for invalid characters by using the GetInvalidPathChars method or path is prefixed with, or contains, only a colon character (:).", ex)
                    Throw ex2

                Catch ex As PlatformNotSupportedException
                    Dim ex2 As New Exception("The specified path, file name, or both exceed the system-defined maximum length. For example, on Windows-based platforms, paths must be less than 248 characters and file names must be less than 260 characters.", ex)
                    Throw ex2

                Catch ex As NotSupportedException
                    Dim ex2 As New Exception("path contains a colon character (:) that is not part of a drive label (""C:\"").", ex)
                    Throw ex2

                Catch ex As Exception
                    Dim ex2 As New Exception("Failed to create directory.")
                    ex2.Data("Directory") = destPath
                    Throw ex2
                End Try
            End If

            For Each file1 As String In IO.Directory.GetFiles(sourcePath)
                Dim dest As String = IO.Path.Combine(destPath, IO.Path.GetFileName(file1))
                If IO.File.Exists(file1) Then
                    If String.Compare("FISLibraryXML.xml", IO.Path.GetFileName(file1)) = 0 Then
                        If Not IO.File.Exists(dest) Then
                            IO.File.Copy(file1, dest, True)
                        End If
                    Else
                        IO.File.Copy(file1, dest, True)  ' Added True here to force the an overwrite 
                    End If
                End If
            Next

            ' Use directly the sourcePath passed in, not the parent of that path
            For Each dir1 As String In IO.Directory.GetDirectories(sourcePath)
                Dim destdir As String = IO.Path.Combine(destPath, IO.Path.GetFileName(dir1))
                CopyDirectory(dir1, destdir)
            Next
        End Sub

    End Class

End Namespace

File: GCDCore/Project/ProjectManagerUI.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Data;
using System.Diagnostics;
using System.IO;

namespace GCDCore.Project
{
    public class ProjectManagerUI : ProjectManagerBase
    {
        protected static RasterPyramidManager m_PyramidManager;

        private static FileInfo m_FISLibrary;
        public static bool IsArcMap
        {
            get { return System.Reflection.Assembly.GetEntryAssembly().FullName.ToLower().Contains("arcmap"); }
        }

        public static RasterPyramidManager PyramidManager { get; }

        public static List<ErrorCalculation.FIS.FISLibraryItem> FISLibrary
        {
            get
            {
                if (m_FISLibrary.Exists)
                {
                    return ErrorCalculation.FIS.FISLibraryItem.Load(m_FISLibrary);
                }
                else
                {
                    return new List<ErrorCalculation.FIS.FISLibraryItem>();
                }
            }
            set { ErrorCalculation.FIS.FISLibraryItem.Save(m_FISLibrary, value); }
        }

        private static string ApplicationFolder
        {
            get { return Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "GCD"); }
        }

        public ProjectManagerUI(GCDConsoleLib.Raster.RasterDriver eDefaultRasterType, string sAutomaticPyramids) : base(ApplicationFolder, Properties.Settings.Default.Erosion, Properties.Settings.Default.Deposition, eDefaultRasterType)
        {
            m_PyramidManager = new RasterPyramidManager(sAutomaticPyramids);
            m_FISLibrary = new FileInfo(Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "FISLibrary.xml"));
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="bRemoveMissingrasters"></param>
        /// <returns>True if the GCD project has been altered during the validation process</returns>
        public static bool Validate(bool bRemoveMissingrasters = false)
        {
            if (!Properties.Settings.Default.ValidateProjectOnLoad)
            {
                return false;
            }

            bool bEditedGCDFile = false;
            StringBuilder sOutputMessages = new StringBuilder();

            ProjectDS.ProjectRow rProject = CurrentProject;
            if (!(rProject is ProjectDS.ProjectRow))
            {
                return bEditedGCDFile;
            }

            List<ProjectDS.DEMSurveyRow> lItemsToRemove = new List<ProjectDS.DEMSurveyRow>();
            foreach (ProjectDS.DEMSurveyRow rDEM in rProject.GetDEMSurveyRows())
            {
                FileInfo sPath = GetAbsolutePath(rDEM.Source);
                if (!sPath.Exists)
                {
                    lItemsToRemove.Add(rDEM);
                }
            }

            //Pass bEditedGCDFile as a reference argument between these 3 methods 
            bEditedGCDFile = ValidateAssociatedSurfaces(ref sOutputMessages, ref bEditedGCDFile, bRemoveMissingrasters);
            bEditedGCDFile = ValidateErrorSurfaces(ref sOutputMessages, ref bEditedGCDFile, bRemoveMissingrasters);
            bEditedGCDFile = ValidateDoDAnalysisSurfaces(ref rProject, ref sOutputMessages, ref bEditedGCDFile, bRemoveMissingrasters);

            if (bRemoveMissingrasters)
            {
                CheckCopyStatusOfGCDProject(FilePath, bEditedGCDFile, lItemsToRemove.Count);

                foreach (ProjectDS.DEMSurveyRow rDEM in lItemsToRemove)
                {
                    sOutputMessages.Append("Removing DEM Survey '" + rDEM.Name + "' because the source raster is missing or cannot be found.").AppendLine();
                    ds.DEMSurvey.RemoveDEMSurveyRow(rDEM);
                    Debug.Write(sOutputMessages.ToString());
                    bEditedGCDFile = true;
                }
            }

            try
            {
                save();
            }
            catch (Exception ex)
            {
                ex.Data["Removing datasets"] = sOutputMessages;
                throw;
            }

            return bEditedGCDFile;

            // This code cleans the project of empty folders which were not fully deleted due to inability to remove
            // ESRI locks even after the file has successfully been deleted And removed from GCD xml
            CleanProjectFolderStructure(true);

            return bEditedGCDFile;
        }

        private static bool ValidateAssociatedSurfaces(ref StringBuilder sOutputMessages, ref bool bEditedGCDFile, bool bRemoveMissingRasters = false)
        {
            List<ProjectDS.AssociatedSurfaceRow> lItemsToRemove = new List<ProjectDS.AssociatedSurfaceRow>();
            foreach (ProjectDS.AssociatedSurfaceRow rAssoc in ds.AssociatedSurface.Rows)
            {
                dynamic dbTest = (DataRow)rAssoc;
                if (DBNull.Value == dbTest("OriginalSource"))
                {
                    rAssoc.OriginalSource = rAssoc.Source;
                    string sAssociatedSurfaceName = System.IO.Path.GetFileNameWithoutExtension(rAssoc.OriginalSource);
                    rAssoc.Source = GetRelativePath(OutputManager.AssociatedSurfaceRasterPath(rAssoc.DEMSurveyRow.Name, sAssociatedSurfaceName));
                    CheckCopyStatusOfGCDProject(FilePath, bEditedGCDFile, 1);
                    bEditedGCDFile = true;
                    save();
                }

                FileInfo sPath = GetAbsolutePath(rAssoc.Source);
                if (!sPath.Exists)
                {
                    lItemsToRemove.Add(rAssoc);
                }
            }

            if (bRemoveMissingRasters)
            {
                //Check to see if a copy of the .gcd project is necessary
                CheckCopyStatusOfGCDProject(FilePath, bEditedGCDFile, lItemsToRemove.Count);

                foreach (ProjectDS.AssociatedSurfaceRow rAssoc in lItemsToRemove)
                {
                    sOutputMessages.Append("Removing Associated Surface '" + rAssoc.Name + "' because the source raster is missing or cannot be found.").AppendLine();
                    ds.AssociatedSurface.RemoveAssociatedSurfaceRow(rAssoc);
                    Debug.Write(sOutputMessages.ToString());
                    bEditedGCDFile = true;
                }
            }

            return bEditedGCDFile;
        }

        private static bool ValidateErrorSurfaces(ref StringBuilder sOutputMessages, ref bool bEditedGCDFile, bool bRemoveMissingRasters = false)
        {
            List<ProjectDS.ErrorSurfaceRow> lItemsToRemove = new List<ProjectDS.ErrorSurfaceRow>();
            foreach (ProjectDS.ErrorSurfaceRow rError in ds.ErrorSurface.Rows)
            {
                if (rError.IsSourceNull())
                {
                    lItemsToRemove.Add(rError);
                }
                else
                {
                    FileInfo sPath = GetAbsolutePath(rError.Source);
                    if (!sPath.Exists)
                    {
                        lItemsToRemove.Add(rError);
                    }
                }
            }

            if (bRemoveMissingRasters)
            {
                CheckCopyStatusOfGCDProject(FilePath, bEditedGCDFile, lItemsToRemove.Count);

                foreach (ProjectDS.ErrorSurfaceRow rError in lItemsToRemove)
                {
                    string sMessage = "Removing Error Surface '" + rError.Name + "' because the source raster is missing or cannot be found.";
                    ds.ErrorSurface.RemoveErrorSurfaceRow(rError);
                    sOutputMessages.Append(sMessage).AppendLine();
                    Debug.Write(sOutputMessages.ToString());
                    bEditedGCDFile = true;
                }
            }

            return bEditedGCDFile;
        }

        private static bool ValidateDoDAnalysisSurfaces(ref ProjectDS.ProjectRow rProjectRow, ref StringBuilder sOutputMessages, ref bool bEditedGCDFile, bool bRemoveMissingRasters = false)
        {
            List<ProjectDS.DoDsRow> lItemsToRemove = new List<ProjectDS.DoDsRow>();
            foreach (ProjectDS.DoDsRow rDoD in ds.DoDs.Rows)
            {
                FileInfo sPathDoDRaw = GetAbsolutePath(rDoD.RawDoDPath);
                FileInfo sPathDoDThresh = GetAbsolutePath(rDoD.ThreshDoDPath);
                if (!sPathDoDRaw.Exists | !sPathDoDThresh.Exists)
                {
                    lItemsToRemove.Add(rDoD);
                }
            }

            if (bRemoveMissingRasters)
            {
                CheckCopyStatusOfGCDProject(FilePath, bEditedGCDFile, lItemsToRemove.Count);

                foreach (ProjectDS.DoDsRow rDoD in lItemsToRemove)
                {
                    sOutputMessages.Append("Removing DoD Surface '" + rDoD.Name + "' because the source raster is missing or cannot be found.").AppendLine();
                    ds.DoDs.RemoveDoDsRow(rDoD);
                    Debug.Write(sOutputMessages.ToString());
                    bEditedGCDFile = true;
                }
            }

            return bEditedGCDFile;
        }

        private static void CleanProjectFolderStructure(bool bRemoveEmptyFolders = false)
        {
            if (FilePath == null)
            {
                return;
            }

            string sPath = Path.GetDirectoryName(FilePath);
            try
            {
                if (bRemoveEmptyFolders)
                {
                    DeleteEmptyFolders(sPath);
                }
            }
            catch (Exception ex)
            {
                //Do Nothing
            }
        }

        private static string CheckCopyStatusOfGCDProject(string sGCDProjectFilePath, bool bEditedGCDFile, int iItemsToRemove)
        {
            if (string.IsNullOrEmpty(sGCDProjectFilePath))
            {
                return null;
            }

            //If gcd file has not been edited in previous steps
            if (bEditedGCDFile == false)
            {
                //If there are elements of GCD file currently marked for removal

                if (iItemsToRemove > 0)
                {
                    //Have established that gcd file has not been edited, i.e. don't want to make a backup of something that has already changed
                    //and established that there items from this set to be removed from the gcd file therefore make a copy of the .gcd file

                    //Copy GCD file, give backup file timestamp of when copy is made
                    string sGCDProjectCopyName = System.IO.Path.GetFileNameWithoutExtension(sGCDProjectFilePath);
                    DateTime sTimeStamp = DateTime.Now;
                    string format = "yyyydMHHmmss";
                    // format is one integer of year day hours minutes seconds
                    sGCDProjectCopyName = sGCDProjectCopyName + sTimeStamp.ToString(format) + ".gcd";
                    string sBackUpPath = Path.Combine(System.IO.Path.GetDirectoryName(sGCDProjectFilePath), "Backup");
                    sGCDProjectCopyName = System.IO.Path.Combine(sBackUpPath, sGCDProjectCopyName);

                    if (System.IO.Directory.Exists(sBackUpPath))
                    {
                        System.IO.File.Copy(sGCDProjectFilePath, sGCDProjectCopyName);
                    }
                    else
                    {
                        System.IO.Directory.CreateDirectory(sBackUpPath);
                        System.IO.File.Copy(sGCDProjectFilePath, sGCDProjectCopyName);
                    }

                    return sGCDProjectCopyName;
                }
            }

            return null;
        }

        private static void DeleteEmptyFolders(string sPath)
        {
            string[] SubDirectories = Directory.GetDirectories(sPath);
            foreach (string strDirectory in SubDirectories)
            {
                DeleteEmptyFolders(strDirectory);
            }

            if (Directory.GetFiles(sPath).Length + Directory.GetDirectories(sPath).Length == 0)
            {
                Directory.Delete(sPath);
                string sMessage = "Deleting folder '" + sPath + "' from project file system because the folder is empty." + Environment.NewLine;
                Debug.Write(sMessage.ToString());
            }
        }

        public static void CopyDeployFolder()
        {
            string sDestinationFolder = ApplicationFolder;

            //New Code to test this may not be what you intend though
            string sFolderName = "Deploy";

            if (string.Compare(sFolderName, "Reources\\FIS", true) == 0)
            {
                sDestinationFolder = Path.Combine(sDestinationFolder, sFolderName);
                dynamic sExecutingAssemblyFolder = System.IO.Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location);
                string sOriginalFolder = Path.Combine(sExecutingAssemblyFolder, sFolderName);

                if (Directory.Exists(Path.GetDirectoryName(sDestinationFolder)))
                {
                    CopyDirectory(sOriginalFolder, sDestinationFolder);
                }
                else if (!Directory.Exists(Path.GetDirectoryName(sDestinationFolder)))
                {
                    Directory.CreateDirectory(sDestinationFolder);
                    Debug.WriteLine("Copying AddIn Folder \"" + sOriginalFolder + "\" to \"" + sDestinationFolder + "\"");
                    CopyDirectory(sOriginalFolder, sDestinationFolder);
                }
            }
            else
            {
                Directory.CreateDirectory(sDestinationFolder);
                string sOriginalFolder = Path.Combine(Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location), "Deploy");
                //sDestinationFolder = Path.Combine(sDestinationFolder, Path.GetFileNameWithoutExtension(sFolderName))
                Debug.WriteLine("Copying AddIn Folder \"" + sOriginalFolder + "\" to \"" + sDestinationFolder + "\"");
                CopyDirectory(sOriginalFolder, ApplicationFolder);
            }
        }

        private static void CopyDirectory(string sourcePath, string destPath)
        {
            if (!Directory.Exists(destPath))
            {
                try
                {
                    DirectoryInfo dFolder = Directory.CreateDirectory(destPath);
                    if (!dFolder.Exists)
                    {
                        Exception ex = new Exception("Failed to create directory.");
                        ex.Data["Directory"] = destPath;
                        throw ex;
                    }
                }
                catch (DirectoryNotFoundException ex)
                {
                    Exception ex2 = new Exception("The specified path is invalid (for example, it is on an unmapped drive).", ex);
                    throw ex2;
                }
                catch (IOException ex)
                {
                    Exception ex2 = new Exception("The directory specified by path is a file or the network name is not known.", ex);
                    throw ex2;
                }
                catch (UnauthorizedAccessException ex)
                {
                    Exception ex2 = new Exception("The caller does not have the required permission.", ex);
                    throw ex2;
                }
                catch (ArgumentNullException ex)
                {
                    Exception ex2 = new Exception("Path is Nothing.", ex);
                    throw ex2;
                }
                catch (ArgumentException ex)
                {
                    Exception ex2 = new Exception("The path is a zero-length string, contains only white space, or contains one or more invalid characters. You can query for invalid characters by using the GetInvalidPathChars method or path is prefixed with, or contains, only a colon character (:).", ex);
                    throw ex2;
                }
                catch (PlatformNotSupportedException ex)
                {
                    Exception ex2 = new Exception("The specified path, file name, or both exceed the system-defined maximum length. For example, on Windows-based platforms, paths must be less than 248 characters and file names must be less than 260 characters.", ex);
                    throw ex2;
                }
                catch (NotSupportedException ex)
                {
                    Exception ex2 = new Exception("path contains a colon character (:) that is not part of a drive label (\"C:\\\").", ex);
                    throw ex2;
                }
                catch (Exception ex)
                {
                    Exception ex2 = new Exception("Failed to create directory.");
                    ex2.Data["Directory"] = destPath;
                    throw ex2;
                }
            }

            foreach (string file1 in Directory.GetFiles(sourcePath))
            {
                string dest = Path.Combine(destPath, Path.GetFileName(file1));
                if (File.Exists(file1))
                {
                    if (string.Compare("FISLibraryXML.xml", Path.GetFileName(file1)) == 0)
                    {
                        if (!File.Exists(dest))
                        {
                            File.Copy(file1, dest, true);
                        }
                    }
                    else
                    {
                        File.Copy(file1, dest, true);
                        // Added True here to force the an overwrite 
                    }
                }
            }

            // Use directly the sourcePath passed in, not the parent of that path
            foreach (string dir1 in Directory.GetDirectories(sourcePath))
            {
                string destdir = Path.Combine(destPath, Path.GetFileName(dir1));
                CopyDirectory(dir1, destdir);
            }
        }
    }
}